src/cargo/core/resolver/features.rs

```diff
@@ -0,0 +1,504 @@
+    activated_features: ActivateMap,
+    /// Keeps track of which packages have had its dependencies processed.
+    /// Used to avoid cycles, and to speed up processing.
+    processed_deps: HashSet<(PackageId, bool)>,
+}
+
+impl<'a, 'cfg> FeatureResolver<'a, 'cfg> {
+    /// Runs the resolution algorithm and returns a new `ResolvedFeatures`
+    /// with the result.
+    pub fn resolve(
+        ws: &Workspace<'cfg>,
+        target_data: &RustcTargetData,
+        resolve: &Resolve,
+        requested_features: &RequestedFeatures,
+        specs: &[PackageIdSpec],
+        requested_target: CompileKind,
+        has_dev_units: bool,
+    ) -> CargoResult<ResolvedFeatures> {
+        use crate::util::profile;
+        let _p = profile::start("resolve features");
+
+        let opts = FeatureOpts::new(ws.config(), has_dev_units)?;
+        if !opts.new_resolver {
+            // Legacy mode.
+            return Ok(ResolvedFeatures {
+                activated_features: HashMap::new(),
+                legacy: Some(resolve.features_clone()),
+                opts,
+            });
+        }
+        let mut r = FeatureResolver {
+            ws,
+            target_data,
+            requested_target,
+            resolve,
+            opts,
+            activated_features: HashMap::new(),
+            processed_deps: HashSet::new(),
+        };
+        r.do_resolve(specs, requested_features)?;
+        log::debug!("features={:#?}", r.activated_features);
+        if r.opts.compare {
+            r.compare();
+        }
+        Ok(ResolvedFeatures {
+            activated_features: r.activated_features,
+            legacy: None,
+            opts: r.opts,
+        })
+    }
+
+    /// Performs the process of resolving all features for the resolve graph.
+    fn do_resolve(
+        &mut self,
+        specs: &[PackageIdSpec],
+        requested_features: &RequestedFeatures,
+    ) -> CargoResult<()> {
+        let member_features = self.ws.members_with_features(specs, requested_features)?;
+        for (member, requested_features) in &member_features {
+            let fvs = self.fvs_from_requested(member.package_id(), requested_features);
+            self.activate_pkg(member.package_id(), &fvs, false)?;
+        }
+        Ok(())
+    }
+
+    fn activate_pkg(
+        &mut self,
+        pkg_id: PackageId,
+        fvs: &[FeatureValue],
+        for_build: bool,
+    ) -> CargoResult<()> {
+        // Add an empty entry to ensure everything is covered. This is intended for
+        // finding bugs where the resolver missed something it should have visited.
+        // Remove this in the future if `activated_features` uses an empty default.
+        self.activated_features
+            .entry((pkg_id, for_build))
+            .or_insert_with(BTreeSet::new);
+        for fv in fvs {
+            self.activate_fv(pkg_id, fv, for_build)?;
+        }
+        if !self.processed_deps.insert((pkg_id, for_build)) {
```

Comment:
This guard seems a bit suspicious to me. Just because we visited a package before here doesn't mean we didn't later add features which themselves add dependencies, right?

Reply:
I'm thinking of something like `cargo build -p deep-dependency -p root-dependency` where it activates `deep-dependency` quickly but by activating `root-dependency` we transitively enable more features of `deep-dependency`

Reply:
I think it is fine.  After resolving deep-dependency once, other packages that add features to that deep-dependency will do so via `FeatureValue::CrateFeature` in `activate_fv`.  This in turn checks if it is activating an optional dependency, and will recurse into that if necessary.

This is kinda fundamental to the way this is organized.  If you notice, a few lines below it *always* skips optional dependencies. Whenever a dependency is encountered that enables an optional feature, it will enable it and recurse right away. 

I believe this is structured somewhat differently from how the dependency resolver works.  Presumably because it is figuring out the features as it goes along, and doesn't have the full graph built, yet.

Reply:
Ah ok the skipping of optional deps below actually does make this sense yeah, sorry I missed that! It still feels a bit funky to me but because it's doing the skip I'm fine deferring to tests to verify the correctness of this. Mind adding a comment here clarifying what sort of recursion this is guarding against, and how more recursion is actually allowed via feature activation below in a limited way?